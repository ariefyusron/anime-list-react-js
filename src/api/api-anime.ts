import { apiAnime } from ".";
import { ParamsGetAnimeList, ResponseGetAnimeList } from "../types/api-anime";

const listApi = {
  getAnimeList: (params: ParamsGetAnimeList): Promise<ResponseGetAnimeList> => {
    const queryString = new URLSearchParams(Object.entries(params));
    return apiAnime.get(`anime?${queryString.toString()}`);
  }
};

export default listApi;
