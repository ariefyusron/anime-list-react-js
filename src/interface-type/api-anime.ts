export interface ParamsGetAnimeList {
  page?: number;
  limit?: number;
  q?: string;
}

interface Pagination {
  last_visible_page: number;
  has_next_page: boolean;
  current_page: number;
  items: {
    count: number;
    total: number;
    per_page: number;
  };
}

interface ImageSet {
  image_url: string;
  small_image_url: string;
  large_image_url: string;
}

interface Trailer {
  youtube_id: string;
  url: string;
  embed_url: string;
  images: {
    image_url: string;
    small_image_url: string;
    medium_image_url: string;
    large_image_url: string;
    maximum_image_url: string;
  };
}

interface Title {
  type: string;
  title: string;
}

interface Aired {
  from: string;
  to: string;
  prop: {
    from: { day: number; month: number; year: number };
    to: { day: number; month: number; year: number };
  };
  string: string;
}

interface Producer {
  mal_id: number;
  type: string;
  name: string;
  url: string;
}

interface Genre {
  mal_id: number;
  type: string;
  name: string;
  url: string;
}

interface Broadcast {
  day: string;
  time: string;
  timezone: string;
  string: string;
}

interface Data {
  mal_id: number;
  url: string;
  images: {
    jpg: ImageSet;
    webp: ImageSet;
  };
  trailer: Trailer;
  approved: boolean;
  titles: Title[];
  title: string;
  title_english: string;
  title_japanese: string;
  title_synonyms: string[];
  type: string;
  source: string;
  episodes: number;
  status: string;
  airing: boolean;
  aired: Aired;
  duration: string;
  rating: string;
  score: number;
  scored_by: number;
  rank: number;
  popularity: number;
  members: number;
  favorites: number;
  synopsis: string;
  background: string;
  season: string;
  year: number;
  broadcast: Broadcast;
  producers: Producer[];
  licensors: Producer[];
  studios: Producer[];
  genres: Genre[];
  explicit_genres: Genre[];
  themes: Genre[];
  demographics: Genre[];
}

export interface ResponseGetAnimeList {
  pagination: Pagination;
  data: Data[];
}
