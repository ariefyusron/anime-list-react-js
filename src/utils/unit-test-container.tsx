import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { ReactNode } from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import themes from "@themes";

const queryClient = new QueryClient();

const renderWithTheme = (ui: ReactNode) => {
  return render(
    <ThemeProvider theme={themes}>
      <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
    </ThemeProvider>
  );
};

export { renderWithTheme };
