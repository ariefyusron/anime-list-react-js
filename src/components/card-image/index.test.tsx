import { screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";

import { ResponseGetAnimeList } from "@interface-type/api-anime";
import { renderWithTheme } from "@utils/unit-test-container";

import CardImage from ".";

const data = {
  mal_id: 1,
  images: {
    webp: {
      image_url: "link.url"
    }
  },
  title: "Title",
  genres: [
    {
      name: "Action"
    },
    {
      name: "Comedy"
    }
  ],
  duration: "10 min",
  episodes: 10
} as ResponseGetAnimeList["data"][0];

const mockNavigate = vi.fn();

describe("CardImage Component", () => {
  vi.mock("react-router", () => ({
    useNavigate: () => mockNavigate
  }));

  beforeEach(() => {
    mockNavigate.mockClear();
  });

  it("should render the CardImage component type TV", () => {
    const component = renderWithTheme(
      <CardImage
        data-testid="component-card-image-container"
        data={{ ...data, type: "TV" }}
      />
    );

    const imageElement = screen.getByAltText("image-anime");
    expect(imageElement).toBeInTheDocument();
    expect(imageElement).toHaveAttribute("src", "link.url");

    const titleElement = screen.getByText("Title");
    expect(titleElement).toBeInTheDocument();

    expect(component).toMatchSnapshot();

    const containerElement = screen.getByTestId(
      "component-card-image-container"
    );
    fireEvent.mouseEnter(containerElement);
    expect(screen.getByText("TV")).toBeInTheDocument();
    expect(screen.getByText("10 episodes")).toBeInTheDocument();
    expect(screen.getByText("Action - Comedy")).toBeInTheDocument();

    expect(component).toMatchSnapshot();

    fireEvent.click(containerElement);
    expect(mockNavigate).toHaveBeenCalledWith("/detail/1");
  });

  it("should render the CardImage component type OVA", () => {
    renderWithTheme(
      <CardImage
        data-testid="component-card-image-container"
        data={{ ...data, type: "OVA" }}
      />
    );

    const containerElement = screen.getByTestId(
      "component-card-image-container"
    );
    fireEvent.mouseEnter(containerElement);
    expect(screen.getByText("OVA")).toBeInTheDocument();
    expect(screen.getByText("10 episodes")).toBeInTheDocument();
  });

  it("should render the CardImage component type Special", () => {
    renderWithTheme(
      <CardImage
        data-testid="component-card-image-container"
        data={{ ...data, type: "Special" }}
      />
    );

    const containerElement = screen.getByTestId(
      "component-card-image-container"
    );
    fireEvent.mouseEnter(containerElement);
    expect(screen.getByText("Special")).toBeInTheDocument();
    expect(screen.getByText("10 episodes")).toBeInTheDocument();
  });

  it("should render the CardImage component type Ona", () => {
    renderWithTheme(
      <CardImage
        data-testid="component-card-image-container"
        data={{ ...data, type: "Ona" }}
      />
    );

    const containerElement = screen.getByTestId(
      "component-card-image-container"
    );
    fireEvent.mouseEnter(containerElement);
    expect(screen.getByText("Ona")).toBeInTheDocument();
    expect(screen.getByText("10 episodes")).toBeInTheDocument();
  });

  it("should render the CardImage component type Movie", () => {
    renderWithTheme(
      <CardImage
        data-testid="component-card-image-container"
        data={{ ...data, type: "Movie" }}
      />
    );

    const containerElement = screen.getByTestId(
      "component-card-image-container"
    );
    fireEvent.mouseEnter(containerElement);
    expect(screen.getByText("Movie")).toBeInTheDocument();
    expect(screen.getByText("10 min")).toBeInTheDocument();
  });

  it("should render the CardImage component type Music", () => {
    renderWithTheme(
      <CardImage
        data-testid="component-card-image-container"
        data={{ ...data, type: "Music" }}
      />
    );

    const containerElement = screen.getByTestId(
      "component-card-image-container"
    );
    fireEvent.mouseEnter(containerElement);
    expect(screen.getByText("Music")).toBeInTheDocument();
    expect(screen.getByText("10 min")).toBeInTheDocument();
  });

  it("should render the CardImage component type CM", () => {
    renderWithTheme(
      <CardImage
        data-testid="component-card-image-container"
        data={{ ...data, type: "CM" }}
      />
    );

    const containerElement = screen.getByTestId(
      "component-card-image-container"
    );
    fireEvent.mouseEnter(containerElement);
    expect(screen.getByText("CM")).toBeInTheDocument();
    expect(screen.getByText("10 min")).toBeInTheDocument();
  });

  it("should render the CardImage component type PV", () => {
    renderWithTheme(
      <CardImage
        data-testid="component-card-image-container"
        data={{ ...data, type: "PV" }}
      />
    );

    const containerElement = screen.getByTestId(
      "component-card-image-container"
    );
    fireEvent.mouseEnter(containerElement);
    expect(screen.getByText("PV")).toBeInTheDocument();
    expect(screen.getByText("10 min")).toBeInTheDocument();
  });

  it("should render the CardImage component type TV Special", () => {
    renderWithTheme(
      <CardImage
        data-testid="component-card-image-container"
        data={{ ...data, type: "TV Special" }}
      />
    );

    const containerElement = screen.getByTestId(
      "component-card-image-container"
    );
    fireEvent.mouseEnter(containerElement);
    expect(screen.getByText("TV Special")).toBeInTheDocument();
    expect(screen.getByText("10 min")).toBeInTheDocument();
  });
});
