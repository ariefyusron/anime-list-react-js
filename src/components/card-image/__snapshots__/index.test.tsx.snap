// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardImage Component > should render the CardImage component type TV 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-FEMpB jgTKpk"
        data-testid="component-card-image-container"
      >
        <img
          alt="image-anime"
          class="sc-gjcoXW jzMIwe"
          src="link.url"
        />
        <div
          class="sc-dVBluf gElMMC"
        >
          <h2
            class="sc-fLDLck dkpSwt"
          >
            Title
          </h2>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-FEMpB jgTKpk"
      data-testid="component-card-image-container"
    >
      <img
        alt="image-anime"
        class="sc-gjcoXW jzMIwe"
        src="link.url"
      />
      <div
        class="sc-dVBluf gElMMC"
      >
        <h2
          class="sc-fLDLck dkpSwt"
        >
          Title
        </h2>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CardImage Component > should render the CardImage component type TV 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-FEMpB jgTKpk"
        data-testid="component-card-image-container"
      >
        <img
          alt="image-anime"
          class="sc-gjcoXW jzMIwe"
          src="link.url"
        />
        <div
          class="sc-dVBluf gElMMC"
        >
          <h2
            class="sc-fLDLck dkpSwt"
            style="text-overflow: unset; -webkit-line-clamp: unset;"
          >
            Title
          </h2>
          <div
            class="sc-dycHyt gzJjCp"
          >
            <p
              class="sc-jbAkgO fAjAJG"
            >
              TV
            </p>
            <p
              class="sc-gNZgCX cDiVFB"
            >
              10 episodes
            </p>
          </div>
          <p
            class="sc-gNZgCX sc-kfeOyU cDiVFB cSA-DGX"
          >
            Action - Comedy
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-FEMpB jgTKpk"
      data-testid="component-card-image-container"
    >
      <img
        alt="image-anime"
        class="sc-gjcoXW jzMIwe"
        src="link.url"
      />
      <div
        class="sc-dVBluf gElMMC"
      >
        <h2
          class="sc-fLDLck dkpSwt"
          style="text-overflow: unset; -webkit-line-clamp: unset;"
        >
          Title
        </h2>
        <div
          class="sc-dycHyt gzJjCp"
        >
          <p
            class="sc-jbAkgO fAjAJG"
          >
            TV
          </p>
          <p
            class="sc-gNZgCX cDiVFB"
          >
            10 episodes
          </p>
        </div>
        <p
          class="sc-gNZgCX sc-kfeOyU cDiVFB cSA-DGX"
        >
          Action - Comedy
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
